contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -742,51 +768,152 @@ have a built-in `nil` value.
 
 ## Defaulting
 
-Default resource values are API version-specific, and they are applied during
-the conversion from API-versioned declarative configuration to internal objects
-representing the desired state (`Spec`) of the resource. Subsequent GETs of the
-resource will include the default values explicitly.
-
-Incorporating the default values into the `Spec` ensures that `Spec` depicts the
-full desired state so that it is easier for the system to determine how to
-achieve the state, and for the user to know what to anticipate.
+In general we want default values to be explicitly represented in our APIs,
+rather than asserting that "unspecified fields get the default behavior".  This
+is important so that:
+ - default values can evolve and change in newer API versions
+ - the stored configuration depicts the full desired state, making it easier
+   for the system to determine how to achieve the state, and for the user to
+   know what to anticipate
+
+There are 3 distinct ways that default values can be applied when creating or
+updating (including patch and apply) a resource:
+
+ 1. static: based on the requested API version and possibly other fields in the
+    resource, fields can be assigned values during the API call.
+ 2. admission control: based on the configured admission controllers and
+    possibly other state in or out of the cluster, fields can be assigned
+    values during the API call.
+ 3. controllers: arbitrary changes (within the bounds of what is allowed) can
+    be made to a resource after the API call has completed.
+
+Some care is required when deciding which mechanism to use and managing the
+semantics.
+
+### Static Defaults
+
+Static default values are specific to each API version.  The default field
+values applied when creating an object with the "v1" API may be different than
+the values applied when using the "v2" API.  In most cases, these values are
+defined as literal values by the API version (e.g. "if this field is not
+specified it defaults to 0").  In some cases, these values may be
+deterministically derived from other fields (e.g. "if this field is not
+specified it defaults to the value of otherField").  In very rare cases, these
+values may be allocated from some pool or determined by some other method (e.g.
+Service's IP and IP-family related fields).
+
+These values are applied synchronously by the API server during the conversion
+from API-versioned declarative configuration to internal objects representing
```

Comment:
Do we want this to apply to both CRD-based and compiled-in types? If so, this could generalize to "defaults are applied after decoding versioned data". Also, these defaults are applied when reading from etcd, not just in the incoming API call.

Reply:
for my understanding, defaults on read only apply when the object is stored in an older version without the new fields. When you read that old object, the new fields get the defaults, right?

Reply:
Correct - all the defaulting logic is "if X is unset, set it to Y".

Jordan, do you know: is it the storage-version API version or the being-used API version that decides defaults when reading?

i.e.

* I write an object which is stored with v1 API
* We add a new "foobar" field with default 0
* We add a v2 API and set foobar default to 1
* I read the stored object with v2

Do I see 0 (v1) or 1 (v2) ?  I'd hope it is the v1 default, otherwise different clients will get different reads, but I am not 100% confident in that.

Reply:
Will update text

Reply:
the version of the serialized data being read always determines the defaults applied

> * I write an object which is stored with v1 API

etcd data is `{apiVersion:v1, ...}`

> * We add a new "foobar" field with default 0

etcd data is still `{apiVersion:v1, ...}`, but whenever it is read from etcd, it gets defaulted to `{apiVersion:v1, foobar:0, ...}` internally

> * We add a v2 API and set foobar default to 1
> * I read the stored object with v2

I assume you mean you read with v2 via the API? If so, the path looks like:

* API server reads from etcd, which still has `{apiVersion:v1, ...}`
* API server applies the v1 default to that data, gets `{apiVersion:v1, foobar:0, ...}` internally
* API server converts to v2 and returns `{apiVersion:v2, foobar:0, ...}` to the user

Reply:
That's what I thought.  Good.